File: app/src/main/java/com/example/usermanagmentapp/ui/main/UserAdapter.java
package com.example.usermanagmentapp.ui.main;

import android.view.LayoutInflater;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.paging.PagingDataAdapter;
import androidx.recyclerview.widget.DiffUtil;

import com.example.usermanagmentapp.R;
import com.example.usermanagmentapp.data.model.User;
import com.example.usermanagmentapp.data.model.UserStatus;
import com.example.usermanagmentapp.databinding.RowUserBinding;

import java.util.Objects;
/**
 * Adapter for the list of users.
 */
public class UserAdapter extends PagingDataAdapter<User, UserViewHolder> {
    final UserClickListener clickListener;

    public UserAdapter(UserClickListener clickListener) {
        super(new DiffCallback());
        this.clickListener = clickListener;
    }

    @NonNull
    @Override
    public UserViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        LayoutInflater layoutInflater = LayoutInflater.from(parent.getContext());
        RowUserBinding inflate = RowUserBinding.inflate(layoutInflater, parent, false);
        return new UserViewHolder(inflate);
    }

    @Override
    public void onBindViewHolder(@NonNull UserViewHolder holder, int position) {
        User user = getItem(position);
        if (user != null) {
            holder.imageView.setImageResource(imageFromStatus(user.status));
            holder.userTextView.setText(user.name);
            holder.itemView.setOnClickListener(v -> clickListener.onClick(user));
        }
    }

    private int imageFromStatus(UserStatus status) {
        if (status == UserStatus.Active) {
            return R.drawable.active_shape;
        } else {
            return R.drawable.inactive_shape;
        }
    }


    private static class DiffCallback extends DiffUtil.ItemCallback<User> {

        @Override
        public boolean areItemsTheSame(@NonNull User oldItem, @NonNull User newItem) {
            return Objects.equals(oldItem.id, newItem.id);
        }

        @Override
        public boolean areContentsTheSame(@NonNull User oldItem, @NonNull User newItem) {
            return oldItem.equals(newItem);
        }
    }

    public interface UserClickListener {
        void onClick(User user);
    }
}

File: app/src/main/java/com/example/usermanagmentapp/ui/main/UserAdapter.kt
package com.example.usermanagmentapp.ui.main

import android.util.Log
import android.view.LayoutInflater
import android.view.ViewGroup
import android.widget.TextView
import androidx.paging.PagingDataAdapter
import androidx.recyclerview.widget.DiffUtil
import com.example.usermanagmentapp.R
import com.example.usermanagmentapp.data.model.User
import com.example.usermanagmentapp.data.model.UserStatus
import com.example.usermanagmentapp.databinding.RowUserBinding

class UserAdapter(private val onClick: ((Int, User) -> Unit)? = null) :
    PagingDataAdapter<User, UserViewHolder>(USER_COMPARATOR) {
    companion object {
        private val USER_COMPARATOR = object : DiffUtil.ItemCallback<User>() {
            override fun areItemsTheSame(oldItem: User, newItem: User): Boolean =
                oldItem.id == newItem.id

            override fun areContentsTheSame(oldItem: User, newItem: User): Boolean =
                oldItem == newItem

        }
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): UserViewHolder {
        val from = LayoutInflater.from(parent.context)
        return UserViewHolder(RowUserBinding.inflate(from, parent, false))
    }


    override fun onBindViewHolder(holder: UserViewHolder, position: Int) {
        val user = getItem(position)
        Log.d("MEGA", "onBindViewHolder: $user")
        user?.let {
            val status = user.status
            holder.userTextView.text = user.name
            holder.imageView.setImageResource(imageFromStatus(status))

            holder.itemView.setOnClickListener {
                onClick?.invoke(position,user)
            }
        }
    }

    private fun imageFromStatus(status: UserStatus): Int {

        return when (status) {
            UserStatus.Active -> R.drawable.active_shape
            UserStatus.Inactive -> R.drawable.inactive_shape
        }

    }

}

